File: fusefs/src/encrypted-file.js
const fs = require('node:fs');
const path = require('node:path');

const { encodeEncryptedFileV1, decodeEncryptedFileV1 } = require('./crypto');
const { newDek, encodeWrappedDekV1, decodeWrappedDekV1 } = require('./dek-store');

function sidecarDekPath(realPath) {
  return `${realPath}.ocpfs.dek`;
}

function assertKey32(name, buf) {
  if (!Buffer.isBuffer(buf) || buf.length !== 32) throw new Error(`${name} must be 32 bytes`);
}

function loadOrCreateDek({ kek, realPath, createIfMissing }) {
  assertKey32('kek', kek);
  const dekPath = sidecarDekPath(realPath);

  try {
    const wrapped = fs.readFileSync(dekPath);
    return decodeWrappedDekV1({ kek, blob: wrapped });
  } catch (e) {
    if (e && e.code === 'ENOENT') {
      if (!createIfMissing) {
        const err = new Error('missing DEK sidecar');
        err.code = 'EACCES';
        throw err;
      }
      const dek = newDek();
      const wrapped = encodeWrappedDekV1({ kek, dek });
      fs.mkdirSync(path.dirname(realPath), { recursive: true, mode: 0o700 });
      fs.writeFileSync(dekPath, wrapped, { mode: 0o600 });
      return dek;
    }
    throw e;
  }
}

function readEncryptedFile({ kek, realPath, createIfMissing = false }) {
  const dek = loadOrCreateDek({ kek, realPath, createIfMissing });

  try {
    const blob = fs.readFileSync(realPath);
    const plaintext = decodeEncryptedFileV1({ dek, blob });
    return { dek, plaintext };
  } catch (e) {
    if (e && e.code === 'ENOENT') return { dek, plaintext: Buffer.alloc(0) };
    const err = new Error('ciphertext decode failed');
    err.code = 'EACCES';
    throw err;
  }
}

function writeEncryptedFile({ dek, realPath, plaintext }) {
  assertKey32('dek', dek);
  if (!Buffer.isBuffer(plaintext)) throw new Error('plaintext must be a Buffer');

  const blob = encodeEncryptedFileV1({ dek, plaintext });
  fs.mkdirSync(path.dirname(realPath), { recursive: true, mode: 0o700 });
  fs.writeFileSync(realPath, blob, { mode: 0o600 });
}

module.exports = {
  sidecarDekPath,
  readEncryptedFile,
  writeEncryptedFile,
};

File: fusefs-swift/Sources/OcProtectFsFuseCore/EncryptedFile.swift
import Foundation

// Phase 3 (Issue #109): encrypted file + DEK sidecar handling parity with fusefs/src/encrypted-file.js.

public enum EncryptedFile {
  public static func sidecarDekPath(_ realPath: String) -> String {
    "\(realPath).ocpfs.dek"
  }

  public struct EncryptedFileError: Error, CustomStringConvertible {
    public let description: String
    public let code: Int32

    public static func eacces(_ msg: String) -> EncryptedFileError {
      EncryptedFileError(description: msg, code: EACCES)
    }
  }

  public static func loadOrCreateDek(kek: Data, realPath: String, createIfMissing: Bool) throws -> Data {
    let dekPath = sidecarDekPath(realPath)

    do {
      let wrapped = try Data(contentsOf: URL(fileURLWithPath: dekPath))
      return try DekStoreV1.decodeWrappedDek(kek: kek, blob: wrapped)
    } catch let e as CocoaError where e.code == .fileReadNoSuchFile {
      if !createIfMissing {
        throw EncryptedFileError.eacces("missing DEK sidecar")
      }

      let dek = DekStoreV1.newDek()
      let wrapped = try DekStoreV1.encodeWrappedDek(kek: kek, dek: dek)

      let dir = (realPath as NSString).deletingLastPathComponent
      try FileManager.default.createDirectory(atPath: dir, withIntermediateDirectories: true, attributes: [FileAttributeKey.posixPermissions: 0o700])

      // Write with 0600. Use atomic write to avoid partial blobs.
      try wrapped.write(to: URL(fileURLWithPath: dekPath), options: .atomic)
      try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: dekPath)

      return dek
    }
  }

  public static func readEncryptedFile(kek: Data, realPath: String, createIfMissing: Bool = false) throws -> (dek: Data, plaintext: Data) {
    let dek = try loadOrCreateDek(kek: kek, realPath: realPath, createIfMissing: createIfMissing)

    do {
      let blob = try Data(contentsOf: URL(fileURLWithPath: realPath))
      let plaintext = try OcfsCrypto.decodeEncryptedFileV1(dek: dek, blob: blob)
      return (dek, plaintext)
    } catch let e as CocoaError where e.code == .fileReadNoSuchFile {
      return (dek, Data())
    } catch {
      // Map decode errors to EACCES (matches JS behavior).
      throw EncryptedFileError.eacces("ciphertext decode failed")
    }
  }

  public static func writeEncryptedFile(dek: Data, realPath: String, plaintext: Data) throws {
    let blob = try OcfsCrypto.encodeEncryptedFileV1(dek: dek, plaintext: plaintext)

    let dir = (realPath as NSString).deletingLastPathComponent
    try FileManager.default.createDirectory(atPath: dir, withIntermediateDirectories: true, attributes: [FileAttributeKey.posixPermissions: 0o700])

    try blob.write(to: URL(fileURLWithPath: realPath), options: .atomic)
    try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: realPath)
  }
}
